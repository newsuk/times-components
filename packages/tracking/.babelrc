{
  "presets": ["flow", "env", "react", "stage-2"]
}
