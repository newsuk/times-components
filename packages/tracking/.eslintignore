**/node_modules/*
**/coverage/*
dist
flow-typed
