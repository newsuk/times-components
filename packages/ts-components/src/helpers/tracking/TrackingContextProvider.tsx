import React, { useContext, useState } from 'react';
import merge from 'lodash.merge';

import { useIntersectionObserver } from '../../utils/intersectObserverHook';

export type TrackingAttributes = { [key: string]: any };
export type TrackingContext = {
  object?: string;
  component?: string;
  action?: string;
  attrs?: TrackingAttributes;
};

export type TrackingContextProps = {
  analyticsStream?: (analyticsEvent: any) => void;
  context?: TrackingContext;
  scrolledEvent?: TrackingContext;
};

export type TrackingContextType = TrackingContextProps & {
  fireAnalyticsEvent?: (analyticsEvent: any) => void;
};

const TrackingContext = React.createContext<TrackingContextType>({
  context: {}
});

type TrackingContextChildren = {
  children:
    | ((
        props: {
          fireAnalyticsEvent: (evt: TrackingContext) => void;
          intersectObserverRef: (ref: HTMLElement | null) => void;
        }
      ) => JSX.Element | JSX.Element[])
    | JSX.Element
    | JSX.Element[];
};

export const useTrackingContext = () => useContext(TrackingContext);

export const TrackingContextProvider = ({
  children,
  analyticsStream,
  context,
  scrolledEvent
}: TrackingContextProps & TrackingContextChildren) => {
  const parentTrackingContext = useContext(TrackingContext);
  const stream =
    analyticsStream !== undefined
      ? analyticsStream
      : parentTrackingContext.analyticsStream;

  const aggregatedContext = merge({}, parentTrackingContext.context, context);

  const fireAnalyticsEvent = (event: TrackingContext) => {
    const aggregatedEvent = merge({}, aggregatedContext, event, {
      attrs: {
        eventTime: new Date().toISOString()
      }
    });

    stream
      ? stream(aggregatedEvent)
      : // tslint:disable-next-line:no-console
        console.error('no analytics stream to send to', aggregatedEvent);
  };

  const [ref, setRef] = useState<HTMLElement | null>(null);

  if (scrolledEvent) {
    useIntersectionObserver(
      ref,
      () =>
        fireAnalyticsEvent &&
        fireAnalyticsEvent({ action: 'Scrolled', ...scrolledEvent }),
      {
        threshold: 0.5
      }
    );
  }

  return (
    <TrackingContext.Provider
      value={{
        fireAnalyticsEvent,
        context: aggregatedContext,
        analyticsStream: stream
      }}
    >
      {typeof children === 'function'
        ? children({ fireAnalyticsEvent, intersectObserverRef: setRef })
        : children}
    </TrackingContext.Provider>
  );
};
