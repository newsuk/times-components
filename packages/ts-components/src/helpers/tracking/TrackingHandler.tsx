declare global {
  interface Window {
    utag?: {
      link: (params: Record<string, any>) => void;
    };
  }
}

type TrackingParams = Record<string, any>;

export const tealiumTrackingHandler = (
  eventName: string,
  action: string,
  browsingMethod: string,
  additionalParams: TrackingParams = {}
): void => {
  if (window.utag && window.utag.link) {
    window.utag.link({
      event_navigation_name: eventName,
      event_navigation_action: action,
      event_navigation_browsing_method: browsingMethod,
      ...additionalParams
    });
  } else {
    // tslint:disable-next-line:no-console
    console.error('Tealium tracking is not available.');
  }
};
