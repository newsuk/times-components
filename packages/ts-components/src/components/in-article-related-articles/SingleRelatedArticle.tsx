import React from 'react';
import DatePublication from '@times-components/date-publication';

import {
  SingleRelatedArticleContainer,
  SingleRelatedArticlesImageContainer
} from './SingleRelatedArticle.styles';
import { AspectRatio } from '../aspect-ratio/AspectRatio';
import { RelatedArticleType } from './RelatedArticle';
import { useTrackingContext } from '../../helpers/tracking/TrackingContextProvider';
import { handleClick } from './tracking-helpers';

type RelatedArticleProps = {
  sectionColour: string;
} & RelatedArticleType;

export const SingleRelatedArticle = ({
  headline,
  link,
  image,
  summary,
  publishedTime,
  byline
}: RelatedArticleProps) => {
  const { fireAnalyticsEvent } = useTrackingContext();
  return (
    <SingleRelatedArticleContainer>
      {image && (
        <SingleRelatedArticlesImageContainer>
          <a
            href={link}
            onClick={() => handleClick(fireAnalyticsEvent, 'image', headline)}
          >
            <AspectRatio ratio="16:9">
              <img src={image} />
            </AspectRatio>
          </a>
        </SingleRelatedArticlesImageContainer>
      )}
      <section>
        <a
          href={link}
          onClick={() => handleClick(fireAnalyticsEvent, 'headline', headline)}
        >
          <h3>{headline}</h3>
        </a>
        <div className="summary">{summary}</div>
        <div className="publishedTime">
          <DatePublication date={publishedTime!} showDay={false} />
        </div>
        <div className="byline">{byline}</div>
      </section>
    </SingleRelatedArticleContainer>
  );
};
