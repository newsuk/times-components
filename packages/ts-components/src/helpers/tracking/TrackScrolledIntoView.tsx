import React, { useState } from 'react';

import { useTrackingContext } from './TrackingContextProvider';
import { useIntersectionObserver } from '../../utils/intersectObserverHook';
type TrackScrolledIntoViewProps = {
  analyticsEvent: any;
  children: (
    props: { intersectObserverRef: (ref: HTMLDivElement | null) => void }
  ) => JSX.Element;
};
export const TrackScrolledIntoView = ({
  children,
  analyticsEvent
}: TrackScrolledIntoViewProps) => {
  const [ref, setRef] = useState<HTMLDivElement | null>(null);
  const { fireAnalyticsEvent } = useTrackingContext();

  useIntersectionObserver(
    ref,
    () => fireAnalyticsEvent && fireAnalyticsEvent(analyticsEvent),
    {
      threshold: 0.5
    }
  );
  return <>{children({ intersectObserverRef: setRef })}</>;
};
