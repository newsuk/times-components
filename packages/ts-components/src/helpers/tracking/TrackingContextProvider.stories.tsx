import React from 'react';

import { showcaseConverter } from '@times-components/storybook';

import { TrackingContextProvider } from './TrackingContextProvider';

import analyticsStream from '../../fixtures/analytics-actions/analytics-actions';

const showcase = {
  children: [
    {
      component: () => {
        return (
          <TrackingContextProvider
            context={{
              attrs: {
                articleHeadline: 'articleHeadline',
                section: 'section'
              }
            }}
            analyticsStream={analyticsStream}
          >
            {({ fireAnalyticsEvent }) => (
              <div>
                <div>
                  <button
                    onClick={() =>
                      fireAnalyticsEvent({
                        component: 'button 1',
                        action: 'click'
                      })
                    }
                  >
                    Click
                  </button>
                </div>
                <TrackingContextProvider
                  context={{
                    attrs: { subsection: 'nested' }
                  }}
                  analyticsStream={analyticsStream}
                >
                  {({ fireAnalyticsEvent: fireNestedAnalyticsEvent }) => (
                    <button
                      onClick={() =>
                        fireNestedAnalyticsEvent({
                          component: 'button nested',
                          action: 'click'
                        })
                      }
                    >
                      Nested Click
                    </button>
                  )}
                </TrackingContextProvider>
              </div>
            )}
          </TrackingContextProvider>
        );
      },
      name: 'Tracking Context Provider',
      platform: 'web',
      type: 'story'
    }
  ],
  name: 'Typescript Component/Helpers'
};

showcaseConverter(module, showcase);
