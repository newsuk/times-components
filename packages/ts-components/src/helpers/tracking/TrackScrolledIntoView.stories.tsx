import React from 'react';

import { showcaseConverter } from '@times-components/storybook';

import { action } from '@storybook/addon-actions';

import styled from 'styled-components';
import { TrackScrolledIntoView } from './TrackScrolledIntoView';
import { TrackingContextProvider } from './TrackingContextProvider';

const Component = styled.div`
  align-items: center;
  background: #aaa;
  border: solid 5px #555;
  display: flex;
  height: 200px;
  justify-content: center;
  width: 200px;
`;

const StripedContainer = styled.div`
  background: linear-gradient(#f4f4f4, #f4f4f4 50%, #eee 50%, #eee);
  background-size: 100% 20px;
  border: 30px solid #f4f4f4;
  height: 50vh;
  &.full {
    height: 100vh;
  }
`;

const analyticsStream = (event: any) => {
  // tslint:disable-next-line:no-console
  console.log('analytics-action', event);
  action('analytics-action')(event);
};

const scrollEvent = ({ headline }: any) => ({
  component_name: `${headline}`,
  component_type: 'interactive',
  event_navigation_action: 'navigation',
  event_navigation_name: 'in article puff cta clicked',
  event_navigation_browsing_method: 'click'
});

const showcase = {
  children: [
    {
      component: ({ boolean: bool, select }: any) => {
        const show = bool('show', true);
        const height = select(
          'Content size above component',
          {
            'Half Viewport': 'half',
            'Full Viewport': 'full'
          },
          'half'
        );
        return (
          show && (
            <>
              <TrackingContextProvider
                context={{
                  articleHeadline: 'articleHeadline',
                  section: 'section'
                }}
                analyticsStream={analyticsStream}
              >
                <StripedContainer className={height} />
                <TrackScrolledIntoView
                  analyticsEvent={scrollEvent({
                    headline: 'headline1'
                  })}
                >
                  {({ intersectObserverRef }) => (
                    <Component ref={intersectObserverRef}>Wrapped 1</Component>
                  )}
                </TrackScrolledIntoView>
                <StripedContainer className={height} />
                <TrackingContextProvider
                  context={{
                    section: 'a different section'
                  }}
                >
                  <TrackScrolledIntoView
                    analyticsEvent={scrollEvent({
                      headline: 'headline2'
                    })}
                  >
                    {({ intersectObserverRef }) => (
                      <Component ref={intersectObserverRef}>
                        Wrapped 2
                      </Component>
                    )}
                  </TrackScrolledIntoView>
                </TrackingContextProvider>
                <StripedContainer className={height} />
              </TrackingContextProvider>
            </>
          )
        );
      },
      name: 'Basic',
      platform: 'web',
      type: 'story'
    }
  ],
  name: 'Typescript Component/Track Scrolled into View'
};

// @ts-ignore
showcaseConverter(module, showcase);
