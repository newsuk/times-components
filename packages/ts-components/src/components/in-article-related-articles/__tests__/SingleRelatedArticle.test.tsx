import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { SingleRelatedArticle } from '../SingleRelatedArticle';
import { TrackingContextProvider } from '../../../helpers/tracking/TrackingContextProvider';
import mockDate from 'mockdate';

const article1 = {
  label: 'Environment',
  headline: 'Scientists discover ‘chocolate frogs’ in New Guinea',
  link:
    'https://www.thetimes.co.uk/article/scientists-discover-chocolate-frogs-in-new-guinea-bt3z986bj',
  image:
    'https://www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2F464a9dfe-c453-11eb-8601-6a2ece3e4634.jpg?crop=3586%2C2017%2C0%2C187&resize=663',
  summary:
    'Popularised by the Harry Potter books, the chocolate frog, it has emerged, is far from fictional. A living version of the confectionery shop favourite has been discovered in the jungles of New Guinea.',
  publishedTime: '2021-06-03T11:00:00.000Z',
  byline: 'Roger Maynard, Sydney'
};

describe('<SingleRelatedArticle>', () => {
  const realIntl = Intl;

  beforeEach(() => {
    global.Intl = {
      DateTimeFormat: () => ({
        // @ts-ignore
        resolvedOptions: () => ({ timeZone: 'Europe/London' })
      })
    };
  });

  afterEach(() => {
    global.Intl = realIntl;
  });

  it('renders', () => {
    const { baseElement } = render(
      <SingleRelatedArticle sectionColour="red" {...article1} />
    );
    expect(baseElement).toMatchSnapshot();
  });
  it('renders no image', () => {
    const { baseElement } = render(
      <SingleRelatedArticle
        sectionColour="red"
        {...article1}
        image={undefined}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});

describe('tracking', () => {
  beforeEach(() => {
    mockDate.set(1620000000000);
  });

  afterEach(() => {
    mockDate.reset();
  });
  it('fires click event when the image is clicked', () => {
    const analyticsStream = jest.fn();
    const { getByRole } = render(
      <TrackingContextProvider
        context={{
          component: 'ArticleSkeleton',
          attrs: {
            articleHeadline: 'articleHeadline',
            section: 'section'
          }
        }}
        analyticsStream={analyticsStream}
      >
        <SingleRelatedArticle sectionColour="red" {...article1} />
      </TrackingContextProvider>
    );
    fireEvent.click(getByRole('img'));
    expect(analyticsStream).toHaveBeenCalledTimes(1);
    expect(analyticsStream).toHaveBeenCalledWith({
      action: 'Clicked',
      component: 'ArticleSkeleton',
      object: 'InArticleRelatedArticles',
      attrs: {
        articleHeadline: 'articleHeadline',
        component_name:
          'related article : Scientists discover ‘chocolate frogs’ in New Guinea',
        eventTime: '2021-05-03T00:00:00.000Z',
        event_navigation_browsing_method: 'click',
        event_navigation_name:
          'button : image : Scientists discover ‘chocolate frogs’ in New Guinea',
        section: 'section'
      }
    });
  });
  it('fires click event when the headline is clicked', () => {
    const analyticsStream = jest.fn();
    const { getByText } = render(
      <TrackingContextProvider
        context={{
          component: 'ArticleSkeleton',
          attrs: {
            articleHeadline: 'articleHeadline',
            section: 'section'
          }
        }}
        analyticsStream={analyticsStream}
      >
        <SingleRelatedArticle sectionColour="red" {...article1} />
      </TrackingContextProvider>
    );
    fireEvent.click(
      getByText('Scientists discover ‘chocolate frogs’ in New Guinea')
    );
    expect(analyticsStream).toHaveBeenCalledTimes(1);
    expect(analyticsStream).toHaveBeenCalledWith({
      action: 'Clicked',
      component: 'ArticleSkeleton',
      object: 'InArticleRelatedArticles',
      attrs: {
        articleHeadline: 'articleHeadline',
        component_name:
          'related article : Scientists discover ‘chocolate frogs’ in New Guinea',
        eventTime: '2021-05-03T00:00:00.000Z',
        event_navigation_browsing_method: 'click',
        event_navigation_name:
          'button : headline : Scientists discover ‘chocolate frogs’ in New Guinea',
        section: 'section'
      }
    });
  });
});
