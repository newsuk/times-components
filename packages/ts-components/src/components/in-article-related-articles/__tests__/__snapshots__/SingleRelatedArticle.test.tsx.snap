// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleRelatedArticle> renders 1`] = `
<body>
  <div>
    <div
      class="sc-bdVaJa jfCTOX"
    >
      <div
        class="sc-bwzfXH dFUiIq"
      >
        <a
          href="https://www.thetimes.co.uk/article/scientists-discover-chocolate-frogs-in-new-guinea-bt3z986bj"
        >
          <div
            class="sc-htpNat kbcxtj"
          >
            <img
              src="https://www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2F464a9dfe-c453-11eb-8601-6a2ece3e4634.jpg?crop=3586%2C2017%2C0%2C187&resize=663"
            />
          </div>
        </a>
      </div>
      <section>
        <a
          href="https://www.thetimes.co.uk/article/scientists-discover-chocolate-frogs-in-new-guinea-bt3z986bj"
        >
          <h3>
            Scientists discover ‘chocolate frogs’ in New Guinea
          </h3>
        </a>
        <div
          class="summary"
        >
          Popularised by the Harry Potter books, the chocolate frog, it has emerged, is far from fictional. A living version of the confectionery shop favourite has been discovered in the jungles of New Guinea.
        </div>
        <div
          class="publishedTime"
        >
          <time
            datetime="2021-06-03T11:00:00.000Z"
          >
            June 03 2021, 12.00pm
            
          </time>
        </div>
        <div
          class="byline"
        >
          Roger Maynard, Sydney
        </div>
      </section>
    </div>
  </div>
</body>
`;

exports[`<SingleRelatedArticle> renders no image 1`] = `
<body>
  <div>
    <div
      class="sc-bdVaJa jfCTOX"
    >
      <section>
        <a
          href="https://www.thetimes.co.uk/article/scientists-discover-chocolate-frogs-in-new-guinea-bt3z986bj"
        >
          <h3>
            Scientists discover ‘chocolate frogs’ in New Guinea
          </h3>
        </a>
        <div
          class="summary"
        >
          Popularised by the Harry Potter books, the chocolate frog, it has emerged, is far from fictional. A living version of the confectionery shop favourite has been discovered in the jungles of New Guinea.
        </div>
        <div
          class="publishedTime"
        >
          <time
            datetime="2021-06-03T11:00:00.000Z"
          >
            June 03 2021, 12.00pm
            
          </time>
        </div>
        <div
          class="byline"
        >
          Roger Maynard, Sydney
        </div>
      </section>
    </div>
  </div>
</body>
`;
